docs/source/contributing/python-style-guide.md

```diff
@@ -0,0 +1,492 @@
+# Python Style Guide
+
+This document is a _diff_ - it describes how Numenta's Python guidelines differ from or are unspecified by [PEP-8](http://www.python.org/dev/peps/pep-0008/) and [PEP-257](http://www.python.org/dev/peps/pep-0257/).
+
+It's meant to be succinct.  Generally, we wish to remain close to the best practices of the Python community.
+
+## Python Version
+
+Numenta software requires Python 2.7.
+
+## Pylint
+
+The NuPIC installation should put `pylint` on your path and you should create a symbolic link at `~/.pylintrc` that points to `$NUPIC/py/pylintrc`.  The pylint command enforces many of the style guidelines from this document and PEP8.
+
+## File Layout
+
+### Executables
+
+If your source file is meant to be an executable, then:
+
+make the first line a _shebang_:
+
+
+    #!/usr/bin/env python
+
+and change its mode:
+
+    $ chmod +x
+
+### Legal Notice
+
+At the top of each source file (after the shebang, if any), include the following legal notice:
+
+```python
+# ----------------------------------------------------------------------
+# Numenta Platform for Intelligent Computing (NuPIC)
+# Copyright (C) 2015, Numenta, Inc.  Unless you have an agreement
+# with Numenta, Inc., for a separate license for this software code, the
+# following terms and conditions apply:
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU Affero Public License version 3 as
+# published by the Free Software Foundation.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+# See the GNU Affero Public License for more details.
+#
+# You should have received a copy of the GNU Affero Public License
+# along with this program.  If not, see http://www.gnu.org/licenses.
+#
+# http://numenta.org/licenses/
+# ----------------------------------------------------------------------
+```
+
+In the date, include any years in which the code was modified, e.g. "2006-2009", "2006, 2008, 2010".
```

Comment:
I think we avoid the comma-separated list since it can get very long. I think the recommendation should always be "YYYY" for single year or "YYYY-YYYY" for the first and last year of modification.